core/src/main/java/com/linecorp/armeria/internal/common/circuitbreaker/AbstractCircuitBreakerClientHandler.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2022 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.common.circuitbreaker;
+
+import java.util.concurrent.CompletionStage;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.circuitbreaker.CircuitBreakerAbortException;
+import com.linecorp.armeria.client.circuitbreaker.CircuitBreakerClientHandler;
+import com.linecorp.armeria.client.circuitbreaker.CircuitBreakerDecision;
+import com.linecorp.armeria.client.circuitbreaker.ClientCircuitBreakerGenerator;
+import com.linecorp.armeria.common.Request;
+import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.common.Response;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.util.CompletionActions;
+
+public abstract class AbstractCircuitBreakerClientHandler<CB, I extends Request, O extends Response>
+        implements CircuitBreakerClientHandler<CB, I> {
+
+    private static final Logger logger = LoggerFactory.getLogger(AbstractCircuitBreakerClientHandler.class);
+    private final ClientCircuitBreakerGenerator<CB> mapping;
+
+    protected AbstractCircuitBreakerClientHandler(ClientCircuitBreakerGenerator<CB> mapping) {
```

Comment:
What do you say to creating this handler per request when `AbstractCircuitBreakerClient.execute()` is called rather than creating per client?
It would create more objects but I think we may clean up some implementations and tweaks.

Reply:
In order to handle this cleanly, I think I need to go back to making `AbstractCircuitBreakerClient` generic, and creating separate `CircuitBreakerClient` and `Resilience4jCircuitBreakerClient` implementations.
Let me re-explore this option 😉